### ISSUE TITLE
ofxCv::Calibration should pass output matrix by reference for undistort

### ISSUE THREAD

**Issue by xionluhnis** [kylemcdonald/ofxCv#176]
I struggled for a few hours trying to understand why the result of `undistort` with my calibration data was always an image with no size.

In my case, my `ofApp::update` function was:
```cpp
void ofApp::update() {
    vidGrabber.update();

    if (vidGrabber.isFrameNew()) {
        if (calib.isReady()) {
            cv::Mat grayImage;
            ofxCv::toCv(vidGrabber.getPixels()).copyTo(grayImage);
            calib.undistort(grayImage, rectImage, interpolation);
        }
    }
}
```

However, the declaration of ofxCv::Calibration::undistort for image undistortion (with input and output) is:
```cpp
void undistort(Mat src, Mat dst, int interpolationMode = INTER_NEAREST);
```

and `dst` is passed to remap, which does the remapping and outputs the data in `dst`.
Except that here, the Mat was empty to start with, pointing to nothing, so `remap` created the content and put it in `dst`, which doesn't update the outside `rectImage` because it's been passed as a copy (a special copy of something that points to nothing, if it did point to something initialized, it wouldn't have triggered the problem).

One solution is to only pass fully initialized Mat objects to `undistort`, but we can avoid forcing users to do that by changing the declaration as:

```cpp
void undistort(Mat src, Mat &dst, int interpolationMode = INTER_NEAREST);
```

and in that case, even if you don't initialize the content of your Mat, it will still be populated correctly.

**Comment by hiroyuki** [comment]
Hello xionluhnis,

I am working on my project with this now and just found you don't have to do this here.
_Mat_ has data pointer in it. So even _undistort_ parameter has no reference of original object, it's has same pointer address in it. So data itself will be updated.

If you can't, I think you should make sure you allocated the pixels before _undistort_ function.

**Comment by xionluhnis** [comment]
Thanks hiroyuki, so the problem in my code was that the Mat object was not "allocated" before being passed to undistort.

Is this "allocate before passing output reference" a specified design pattern of the opencv library?
If it is not, then this seems to me like a bad pattern to require allocation, or at least it should be documented somewhere.

Note that if you pass the output argument by reference, then even if it's not allocated, everything is fine. Is there any reason not to pass by reference then?

**Comment by alberto2000** [comment]
Having the same issue here - my image would just stay black e.g. nothing was being done to the (empty, unallocated) mat I passed into undistort()

**Comment by xionluhnis** [comment]
@alberto2000 does allocating the `Mat` object solve your problem? It should.

Even though that should solve the problem, I think it's a design issue so I'll keep it open because I don't see any good reason for not passing the output `Mat` by reference. Passing by reference avoids the need for pre-allocation.

**Comment by alberto2000** [comment]
@xionluhnis yes it does, thanks a lot. I agree, it would be much nicer and robust if you could just pass a reference.

**Comment by hiroyuki** [comment]
@xionluhnis agree. Thanks your comment.
